quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -160,671 +164,714 @@ fn mul_matches(left_ax: &Expression, right_ax: &Expression) -> bool {
-            limit.saturating_sub(1),
-        ),
-
-        // Subtraction with negation
-        (
-            ref left,
-            InfixOperator::Minus,
-            Expression::Prefix(PrefixExpression {
-                operator: PrefixOperator::Minus,
-                ref expression,
-            }),
-        ) => simplify_infix(
-            left,
-            InfixOperator::Plus,
-            expression,
-            limit.saturating_sub(1),
-        ),
-        (
-            // -expression - right => smaller of (-expression) - right & -(expression + right)
-            ref left @ Expression::Prefix(PrefixExpression {
-                operator: PrefixOperator::Minus,
-                ref expression,
-            }),
-            InfixOperator::Minus,
-            ref right,
-        ) => {
-            let original = sub!(left.clone(), right.clone());
-            let new = simplify_prefix(
-                PrefixOperator::Minus,
-                &simplify_infix(
-                    expression,
-                    InfixOperator::Plus,
-                    right,
-                    limit.saturating_sub(2),
-                ),
-                limit.saturating_sub(1),
-            );
-            min_by_key(original, new, size)
-        }
-
-        // Multiplication with negation
-        (
-            Expression::Prefix(PrefixExpression {
-                operator: PrefixOperator::Minus,
-                expression: ref left,
-            }),
-            InfixOperator::Star,
-            Expression::Prefix(PrefixExpression {
-                operator: PrefixOperator::Minus,
-                expression: ref right,
-            }),
-        ) => simplify_infix(left, InfixOperator::Star, right, limit.saturating_sub(1)),
-        (
-            ref left,
-            InfixOperator::Star,
-            ref right @ Expression::Prefix(PrefixExpression {
-                operator: PrefixOperator::Minus,
-                ref expression,
-            }),
-        ) => {
-            let original = mul!(left.clone(), right.clone());
-            let neg_left = simplify_prefix(PrefixOperator::Minus, left, limit.saturating_sub(1));
-            let new = simplify_infix(
-                &neg_left,
+        })
+    } else {
+        // There are … many cases here
+        match (simplify(l, limit - 1), op, simplify(r, limit - 1)) {
+            //----------------------------------------------------------------
+            // First: only diving one deep, pattern matching on the operation
+            // (Constant folding and cancellations, mostly)
+            //----------------------------------------------------------------
+
+            // Addition and Subtraction
+
+            // Adding with zero
+            (Expression::Number(x), InfixOperator::Plus, other)
+            | (other, InfixOperator::Plus, Expression::Number(x))
+                if is_zero(x) =>
+            {
+                simplify(&other, limit - 1)
```

Comment:
Isn't `other` already simplified on line 184?

Reply:
Yes, good point!

Reply:
This was me being overzealous in responding to @mac01021's review.